pkg/openid4ci/models.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright Avast Software. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package openid4ci
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/hyperledger/aries-framework-go/pkg/doc/verifiable"
+)
+
+// Client provides support for OpenID4CI methods.
+type Client struct {
+	clientConfig *ClientConfig
+}
+
+// JWTSigner represents a type that can do JWT signing. (To be defined better later).
+type JWTSigner interface {
+	// Sign is not yet fully defined. TODO: determine final types to use once an
+	// implementation exists in Aries
+	Sign(clientID, issuer, time, nonce string) ([]byte, error)
+}
+
+// ClientConfig represents the configuration for an OpenID4CI client.
+type ClientConfig struct {
+	// Wallet-hosted URI for callback redirect from issuer, after user authorizes issuer.
+	WalletCallbackURI string
+	// The DID of the wallet user. The requested credential will be bound to this DID.
+	UserDID string
+	// The wallet app instance's OIDC client ID.
+	ClientID string
+	// A handler for signing a JWT for proving possession of userDID.
+	// When this creates a JWT, it should add kid, jwk, or x5c field corresponding to a key bound to UserDID.
+	JWTSigner JWTSigner
+}
+
+// Credential represents a credential in an InitiationRequest.
+// It can be used in one of two mutually-exclusive ways:
+//  1. Setting Format and Types to specify the format of the credential to be requested and further parameters
+//     characterising the type of the credential to be requested.
+//  2. Setting just ID to one of the ID values in one of the Supported Credentials objects in the
+//     credentials_supported Credential Issuer metadata parameter.
+type Credential struct {
+	Format string   `json:"format,omitempty"`
+	Types  []string `json:"types,omitempty"`
+	ID     string   `json:"-"`
```

Comment:
why we need ID here?

Reply:
This type has two different JSON forms it can take - take a look at https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#section-5.1

Reply:
And then also take a look at the custom `MarshalJSON` and `UnmarshalJSON` methods I added

Reply:
Gone in the latest spec version - I've updated the code accordingly